src/test/agent_test.rs

```diff
@@ -123,3 +121,120 @@ fn connection_reuse() {
     }
     assert_eq!(resp.status(), 200);
 }
+
+// Send an HTTP response on the TcpStream at a rate of two bytes every 10
+// milliseconds, for a total of 600 bytes.
+fn dribble_body_respond(stream: &mut TcpStream) -> io::Result<()> {
+    let contents = [b'a'; 300];
+    let headers = format!(
+        "HTTP/1.1 200 OK\r\nContent-Length: {}\r\n\r\n",
+        contents.len() * 2
+    );
+    stream.write_all(headers.as_bytes())?;
+    for i in 0..contents.len() {
+        stream.write_all(&contents[i..i + 1])?;
+        stream.write_all(&[b'\n'; 1])?;
+        stream.flush()?;
+        thread::sleep(Duration::from_millis(10));
+    }
+    Ok(())
+}
+
+// Read a stream until reaching a blank line, in order to consume
+// request headers.
+fn read_headers(stream: &TcpStream) {
+    for line in BufReader::new(stream).lines() {
+        let line = match line {
+            Ok(x) => x,
+            Err(_) => return,
+        };
+        if line == "" {
+            break;
+        }
+    }
+}
+
+// Start a test server on an available port, that dribbles out a response at 1 write per 10ms.
+// Return the port this server is listening on.
+fn start_dribble_body_server() -> u16 {
+    let listener = std::net::TcpListener::bind("localhost:0").unwrap();
+    let port = listener.local_addr().unwrap().port();
+    let dribble_handler = |mut stream: TcpStream| {
+        read_headers(&stream);
+        if let Err(e) = dribble_body_respond(&mut stream) {
+            eprintln!("sending dribble repsonse: {}", e);
+        }
+    };
+    thread::spawn(move || {
+        for stream in listener.incoming() {
+            thread::spawn(move || dribble_handler(stream.unwrap()));
+        }
```

Comment:
I guess threads are pretty cheap and we don't really care that this thread will hang around for the entire test suite, right?

Would a tweak be to not have a for loop, and just accept one single connection, then exit?

Reply:
Normally I like to clean up threads when tests are done, but I don't yet know how to do this in Rust in a clean and simple way. I think I'd need to set up a condition variable and a select of some sort between listener.incoming() and the condition variable. But listener.incoming() doesn't seem to have affordances for that. I'm interested in learning how to do that, but for now I decided to just leak the thread until the tests are done.

I considered the one-accept-and-then-done approach but concluded it would be less realistic. For instance, if something triggered a retry during the tests, we would get a failure, but for the wrong reason.

Ultimately I'd like to evolve these test servers into something that can be used for more of the test cases, so developing a clean shutdown mechanism will be worthwhile.

Reply:
> I think I'd need to set up a condition variable and a select of some sort between listener.incoming() and the condition variable.

If you were outside rust, how does one normally do this? `listener.incoming()` is blocking, so I assume it's using some underlying syscall that also blocks. Are there variants of that syscall with a timeout like for read/write?

> I considered the one-accept-and-then-done approach but concluded it would be less realistic. For instance, if something triggered a retry during the tests, we would get a failure, but for the wrong reason.

Fair enough!

> Ultimately I'd like to evolve these test servers into something that can be used for more of the test cases, so developing a clean shutdown mechanism will be worthwhile.

You got plans! 😎 

My attention is mostly on hreq now, and incidentally just this week I've been hacking on a server side to it. Today I wrote the very rows where I also [need a shutdown mechanism](https://github.com/algesten/hreq/blob/server/src/server/mod.rs#L89-L91), though it might be simpler because of async.